cpp/src/arrow/array/util.cc

```diff
@@ -442,9 +442,18 @@ class NullArrayFactory {
     // First buffer is always null
     out_->buffers[0] = nullptr;
 
-    // Type codes are all zero, so we can use buffer_ which has had it's memory
-    // zeroed
     out_->buffers[1] = buffer_;
+    // buffer_ is zeroed, but 0 may not be a valid type code
+    if (type.type_codes()[0] != 0) {
+      std::memset(buffer_->mutable_data(), type.type_codes()[0], buffer_->size());
```

Comment:
Hmm, we should do the converse here: keep `buffer_` zeroed and create a new buffer for type codes.